File: app/src/main/java/io/vinter/trackmyanime/utils/AnimeListRecyclerAdapter.java
package io.vinter.trackmyanime.utils;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;

import java.util.List;

import io.vinter.trackmyanime.R;
import io.vinter.trackmyanime.entity.animelist.AnimeListItem;

public class AnimeListRecyclerAdapter extends RecyclerView.Adapter<AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder>{

    private Context context;
    private java.util.List<AnimeListItem> animeList;
    private AnimeListClickListener listener;

    public AnimeListRecyclerAdapter(Context context, List<AnimeListItem> list, AnimeListClickListener listener){
        this.context = context;
        this.animeList = list;
        this.listener = listener;
    }

    @NonNull
    @Override
    public AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        View itemView = inflater.inflate(R.layout.item_anime_list, viewGroup, false);
        return new AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder holder, int i) {
        GlideApp.with(context)
                .load(animeList.get(i).getImageUrl())
                .override(200, 270)
                .placeholder(R.color.colorInactive)
                .error(R.color.colorInactive)
                .transforms(new CenterCrop(), new RoundedCorners(15))
                .into(holder.art);
        holder.art.setOnClickListener(view -> {
            listener.onAnimeClick(animeList.get(i).getMalId(), 0);
        });
        holder.title.setText(animeList.get(i).getTitle());
        holder.eps.setText(String.valueOf(animeList.get(i).getWatchedEps() + "/" + animeList.get(i).getEps()));
        holder.status.setText(animeList.get(i).getStatus());
        if (animeList.get(i).getStatus().equals("completed")) {
            holder.add.setVisibility(View.GONE);
        } else {
            holder.add.setVisibility(View.VISIBLE);
        }
        holder.add.setOnClickListener(view -> {
            listener.onAnimeClick(animeList.get(i).getMalId(), 1);
        });
        holder.eps.setOnClickListener(view -> {
            listener.onAnimeClick(animeList.get(i).getMalId(), 2);
        });
    }

    public void update(List<AnimeListItem> newItems){
        this.animeList = newItems;
        this.notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return animeList.size();
    }

    static class AnimeListRecyclerViewHolder extends RecyclerView.ViewHolder {

        ImageView art;
        TextView title;
        TextView status;
        TextView eps;
        Button add;

        AnimeListRecyclerViewHolder(@NonNull View itemView) {
            super(itemView);
            art = itemView.findViewById(R.id.item_anime_list_pic);
            title = itemView.findViewById(R.id.item_anime_list_title);
            status = itemView.findViewById(R.id.item_anime_list_status);
            eps = itemView.findViewById(R.id.item_anime_list_episodes);
            add = itemView.findViewById(R.id.item_anime_list_add);
        }
    }
}

File: app/src/main/java/io/vinter/trackmyanime/utils/AnimeListRecyclerAdapter.kt
package io.vinter.trackmyanime.utils

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException

import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target

import io.vinter.trackmyanime.R
import io.vinter.trackmyanime.entity.animelist.AnimeListItem

class AnimeListRecyclerAdapter(private val context: Context, private var animeList: List<AnimeListItem>, private val listener: (Int, Int) -> Unit) : RecyclerView.Adapter<AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder {
        val inflater = LayoutInflater.from(viewGroup.context)
        val itemView = inflater.inflate(R.layout.item_anime_list, viewGroup, false)
        return AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: AnimeListRecyclerAdapter.AnimeListRecyclerViewHolder, i: Int) {
        loadImage(i, holder.art)
        holder.art.setOnClickListener { listener(animeList[i].malId!!, 0) }
        holder.art.setOnLongClickListener {
            listener(animeList[i].malId!!, 4)
            true
        }
        holder.title.text = animeList[i].title
        holder.eps.text = animeList[i].watchedEps.toString() + "/" + animeList[i].eps
        holder.status.text = animeList[i].status
        if (animeList[i].status == "completed") {
            holder.add.visibility = View.GONE
        } else {
            holder.add.visibility = View.VISIBLE
        }
        holder.add.setOnClickListener { listener(animeList[i].malId!!, 1) }
        holder.eps.setOnClickListener { listener(animeList[i].malId!!, 2) }
    }

    fun update(newItems: List<AnimeListItem>) {
        this.animeList = newItems
        this.notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return animeList.size
    }

    class AnimeListRecyclerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var art: ImageView = itemView.findViewById(R.id.item_anime_list_pic)
        var title: TextView = itemView.findViewById(R.id.item_anime_list_title)
        var status: TextView = itemView.findViewById(R.id.item_anime_list_status)
        var eps: TextView = itemView.findViewById(R.id.item_anime_list_episodes)
        var add: Button = itemView.findViewById(R.id.item_anime_list_add)
    }

    private fun loadImage(i: Int, art: ImageView){
        GlideApp.with(context)
                .load(animeList[i].imageUrl)
                .override(200, 270)
                .placeholder(R.color.colorInactive)
                .error(R.color.colorInactive)
                .transforms(CenterCrop(), RoundedCorners(15))
                .into(art)
    }

}
